### PR TITLE
fix(plines): 'linebreak' ignores inline virtual text width

### PR DESCRIPTION
**Pull request by freddiehaddad**
## Problem

When `linebreak` is enabled, a whole word wraps to the next screen line instead of part of it. To decide whether a word fits, only text in the buffer is measured, not inline virtual text (e.g. from LSP inlay hints).

So, a word that looks short enough is kept on the line, and the virtual text next to it end up on two lines at the edge of the screen.

Example in a 26-column window. The line in the buffer is `print(greeting, name)`, with an inlay hint `: string` shown after `name`:

    print(greeting, name: stri
    ng)

The hint is split into `: stri` and `ng`.

## Solution

Count inline virtual text when measuring the word, so the decision uses the width that is actually on screen.

Same example:

    print(greeting,
    name: string)

The word moves down in one piece and the hint stays intact.

Lines without inline virtual text are unaffected.

## Performance Note

Two guards keep this code off the hot path unless both of these conditions are true:

1. `linebreak` is on
2. the line has inline virtual text

Redraw time for a full screen (50 rows, 100 columns). Every line holds 25 words; a "mark" is a 1-cell inline virtual text. Lowest of 5 runs.

| case | master (ms) | this PR (ms) | added cost (usec) |
| - | - | - | - |
| linebreak, no marks | 0.318 | 0.326 | +8 |
| linebreak, 1 mark/line | 0.373 | 0.531 | +158 |
| linebreak, 5 marks/line | 0.400 | 0.598 | +198 |
| linebreak, 25 marks/line | 0.616 | 0.968 | +352 |
| no linebreak, 25 marks | 0.558 | 0.542 | -16 |

The first and last rows are the controls: with `linebreak` off or with no virtual text, the cost is unchanged.

The cost is per _visible line that has inline virtual text._ It does not grow with the size of the buffer (measured in a separate run).

| case | master (ms) | this PR (ms) | added cost (usec) |
| - | - | - | - |
| 400 lines, 50 rows, marks on all | 0.615 | 1.007 | +392 |
| 4000 lines, 50 rows, marks on all | 0.624 | 1.028 | +404 |
| 400 lines, 10 rows, marks on all | 0.146 | 0.241 | +95 |
| 4000 lines, 50 rows, marks on 1 line | 0.328 | 0.368 | +40 |
| 400 lines, 50 rows, no marks | 0.317 | 0.333 | +16 |

10× the buffer changes nothing (+392 vs +404 usec). One fifth of the visible rows cuts it about 4× (+392 to +95 usec). Marks on a single line cost +40 usec. That works out to roughly 8 usec per drawn line carrying 25 hints.

**The nested loop is not a problem.** The inner loop walks one extmark's virtual-text chunks (pretty much 1), and the outer loop advances a shared iterator monotonically. In other words, the scan is O(chars + marks in that word), not O(chars x marks).

### CODE REVIEW

File: src/nvim/drawline.c [hunk 1]
```diff
@@ -2428,8 +2428,10 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, b
           char *p = ptr - (mb_off + 1);
 
           CharsizeArg csarg;
-          // lnum == 0, do not want virtual text to be counted here
-          CSType cstype = init_charsize_arg(&csarg, wp, 0, line);
+          CSType cstype = init_charsize_arg(&csarg, wp, lnum, line);
```
Comment (justinmk) [inline src/nvim/drawline.c]: this silently reverts a9c551e5e38f

potential fix? (AI suggestion idk, seems good though):


"kill the dead marktree seek (restores what a9c551e5e38f bought)":

```diff
/// Like init_charsize_arg(), for a caller that measures the text after `cur` but not the inline
/// virtual text at `cur` itself (see CharsizeArg.skip_cur_text). That mode never reads `iter`, so
/// skip seeding it; only the row is needed, for the 'linebreak' lookahead's own iterator.
CSType init_charsize_arg_skip_cur_text(CharsizeArg *csarg, win_T *wp, linenr_T lnum, char *line)
{
  CSType cstype = init_charsize_arg(csarg, wp, 0, line);
  csarg->skip_cur_text = true;
  if (lnum > 0) {
    csarg->virt_row = lnum - 1;
    cstype = kCharsizeRegular;
  }
  return cstype;
}
-          CSType cstype = init_charsize_arg(&csarg, wp, lnum, line);
-          csarg.skip_cur_text = true;
+          CSType cstype = init_charsize_arg_skip_cur_text(&csarg, wp, lnum, line);
```

File: src/nvim/plines.c [hunk 2]
```diff
@@ -316,6 +341,14 @@ CharSize charsize_regular(CharsizeArg *csarg, char *const cur, colnr_T const vco
       }
     }
 
+    // Inline virtual text widens the word, so count it like the loop above does for one
+    // character. Needs its own iterator, since this looks ahead of "csarg->iter".
+    MarkTreeIter virt_iter[1];
+    bool const has_virt = csarg->virt_row >= 0
+                          && marktree_itr_get_filter(buf->b_marktree, csarg->virt_row, 0,
+                                                     csarg->virt_row + 1, 0, inline_filter,
+                                                     virt_iter);
```
Comment (justinmk) [inline src/nvim/plines.c]: AI finding:

> this re-seeks from column 0 of the row for every word. `marktree_itr_get_filter(buf->b_marktree, csarg->virt_row, 0, ...)` then `inline_virt_text_width()` walks and discards every mark below the word start. That's O(words ×
  marks) per line, not the O(chars + marks in the word) the description implies, and it's not only drawing: `linesize_regular`/`getvcol`/`plines_win_col` walk the same line char by char.

suggestion:

Seek from the word start instead, or when `!skip_cur_text` just copy the already-positioned shared iterator.

```diff
-  MarkTreeIter virt_iter[1];
-  bool const has_virt = csarg->virt_row >= 0
-                        && marktree_itr_get_filter(buf->b_marktree, csarg->virt_row, 0,
-                                                   csarg->virt_row + 1, 0, inline_filter,
-                                                   virt_iter);
+  // Seek on first use, from the word start: the loop below often breaks on its first check
+  // (one-char words, end of line), and seeking from column 0 would rescan the whole line.
+  MarkTreeIter virt_iter[1];
+  int has_virt = csarg->virt_row < 0 ? 0 : -1;  // -1: not sought yet

   colnr_T vcol2 = vcol;
   while (true) {
     ...
     int const col = (int)(s - line);
+    if (has_virt < 0) {
+      has_virt = marktree_itr_get_filter(buf->b_marktree, csarg->virt_row, col,
+                                         csarg->virt_row + 1, 0, inline_filter, virt_iter);
+    }
```

File: src/nvim/plines.c [hunk 3]
```diff
@@ -326,7 +359,10 @@ CharSize charsize_regular(CharsizeArg *csarg, char *const cur, colnr_T const vco
         break;
       }
 
-      vcol2 += win_chartabsize(wp, s, vcol2);
+      int const col = (int)(s - line);
+      int const virt_width = has_virt
+                             ? inline_virt_text_width(buf, wp, virt_iter, csarg->virt_row, col) : 0;
+      vcol2 += win_chartabsize(wp, s, vcol2) + virt_width;
```
Comment (justinmk) [inline src/nvim/plines.c]: > TAB inside the measured word gets the wrong padding. vcol2 += win_chartabsize(wp, s, vcol2) + virt_width; evaluates the tab stop before adding the virt width, so a tab drawn after inline text is off by up to ts-1.

suggestion


```diff
-    vcol2 += win_chartabsize(wp, s, vcol2) + virt_width;
+    // Add the virtual text first: it is drawn in front of the character, so a TAB's stop is
+    // computed at the shifted column (the main loop recomputes tab_size for the same reason).
+    vcol2 += virt_width;
+    vcol2 += win_chartabsize(wp, s, vcol2);
```

File: src/nvim/plines.c [hunk 4]
```diff
@@ -118,6 +119,30 @@ CSType init_charsize_arg(CharsizeArg *csarg, win_T *wp, linenr_T lnum, char *lin
   }
 }
 
+/// Total width of inline virtual text at buffer column "col" of row "row", advancing "iter".
+/// Call with non-decreasing "col", like the walks in charsize_regular().
+static int inline_virt_text_width(buf_T *buf, win_T *wp, MarkTreeIter *iter, int row, int col)
```
Comment (justinmk) [inline src/nvim/plines.c]: probably don't need a `buf` param if it accepts `wp` ?


    buf_T *const buf = wp->w_buffer;

File: src/nvim/plines.h [hunk 5]
```diff
@@ -24,6 +24,7 @@ typedef struct {
                              ///< parts of lines, INT_MIN if not yet calculated.
 
   int virt_row;              ///< Row for virtual text, -1 if no virtual text.
+  bool skip_cur_text;        ///< Don't count inline virtual text at the measured character.
```
Comment (justinmk) [inline src/nvim/plines.h]: /// Also means `iter` is not advanced, so such a CharsizeArg cannot be
    /// reused for a later character.

File: src/nvim/drawline.c (hunk 1 above)
Comment (freddiehaddad) [reply src/nvim/drawline.c]: I should probably have done a git blame on that line to understand it better. Tests showed correctness, but I missed the performance reason behind that call. It looked unusual.

Testing that suggestion.

File: src/nvim/plines.c (hunk 2 above)
Comment (freddiehaddad) [reply src/nvim/plines.c]: I'll look closer at this. I think I tried that and there was an edge case that failed. I need to run some tests again on this.

File: src/nvim/drawline.c (hunk 1 above)
Comment (freddiehaddad) [reply src/nvim/drawline.c]: This was a good catch, by AI. It works.

**Comment by justinmk** [comment]
Thanks for your continued attention to detail!
